### PR TITLE
Add System.Net.WebSocket.Protocol netstandard package.

### LINKED ISSUE #21537: Add WebSocket.CreateFromConnectedStream
Several corefx components pull in the internal [ManagedWebSocket implementation](https://github.com/dotnet/corefx/blob/master/src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs), both System.Net.WebSockets.Client on Unix and System.Net.HttpListener on Unix, and with https://github.com/dotnet/corefx/issues/9503 the plan is to use it in System.Net.WebSockets.Client on Windows 7 as well.  Plus, ASP.NET uses the same implementation, via [source cloning](https://github.com/aspnet/WebSockets/blob/dev/src/Microsoft.AspNetCore.WebSockets/Internal/fx/src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs).  We should stop building this multiple times and instead expose it out through an API that anyone can use to wrap an established stream for a websockets connection.
```C#
public abstract class WebSocket
{
    public static WebSocket CreateFromConnectedStream(
        Stream stream,
        bool isServer,
        string subprotocol,
        TimeSpan keepAliveInterval,
        int receiveBufferSize,
        ArraySegment<byte>? receiveBuffer = null);
    ...
}
```
Related to https://github.com/dotnet/corefx/issues/13663#issuecomment-265879275
Related to https://github.com/dotnet/corefx/issues/21509#issuecomment-310862176

Once this is added, we can stop compiling it into the various projects and just use the exposed API.

Alternative:
WebSocket currently exposes the following API:
```C#
        [EditorBrowsable(EditorBrowsableState.Never)]
        public static WebSocket CreateClientWebSocket(Stream innerStream,
            string subProtocol, int receiveBufferSize, int sendBufferSize,
            TimeSpan keepAliveInterval, bool useZeroMaskingKey, ArraySegment<byte> internalBuffer)
```
We could remove that EditorBrowsable(Never) and add a corresponding CreateServerWebSocket API.  Both would effectively use CreateFromConnectedStream, just with a different value for isServer.  However, it's not clear we could overload the meaning of CreateClientWebSocket in corefx to be what we need, given its current semantics on desktop, in particular as on desktop it uses WPSC and is documented as such.

EDIT 9/26/2017:
```C#
// System.Net.WebSockets.Protocol.dll, with a netstandard2.0 build
// (and subsequently specialized for netcoreapp2.1)
namespace System.Net.WebSockets
{
    public static class WebSocketProtocol
    {
        public static WebSocket CreateFromConnectedStream(
            Stream stream,
            bool isServer,
            string subprotocol,
            TimeSpan keepAliveInterval,
            int receiveBufferSize, int sendBufferSize, // hints as to desired size
            ArraySegment<byte>? buffer = null); // buffer space to use however the implementation wants
    }
}
```

### PR DESCRIPTION
**Pull request by Priya91**
fixes #21537 

cc @karelz @stephentoub @wfurt 

cc @weshaggard @ericstj Can you please take a look at the packaging changes for oob assembly..

Short summary:
--------------------
Split the ```ManagedWebSocket``` class into 3 files: ```ManagedWebSocket.netstandard.cs``` and ```ManagedWebSocket.netcoreapp.cs```. I tried to put the common code in ```ManagedWebSocket.Common.cs```. Refactoring out common parts seemed like a backward effort, where all the Span/Memory enhancements may have to be abandoned. So I prioritized enhancements over common code, where the change was tricky to make. Needless to say, the ```*.netcoreapp.cs``` file contains Span/Memory stuff, and the ```*.netstandard.cs``` contains ArraySegment/byte[] stuff.

Testing:
-----------------
I was not able to get a successful test with ```WebSocket.ReceiveAsync```, it fails in ReceiveAsync, because the response from the server is not in the right format. So the code fails with ```ProtocolError``` in ```ReceiveAsync```. The behavior is same with and without this refactoring. Hence didn't include ```ReceiveAsync``` in the ```WebSocketProtocol_CreateFromConnectedStream_Succeeds``` outerloop test.

### CODE REVIEW

File: src/System.Net.WebSockets.Protocol/dir.props [hunk 1]
```diff
@@ -0,0 +1,8 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="..\dir.props" />
+  <PropertyGroup>
+    <AssemblyVersion>4.0.0.0</AssemblyVersion>
+    <AssemblyKey>MSFT</AssemblyKey>
```
Comment (weshaggard) [inline src/System.Net.WebSockets.Protocol/dir.props]: Should be Open key.

**Comment by weshaggard** [comment]
>cc @weshaggard @ericstj Can you please take a look at the packaging changes for oob assembly..

I don't see any packaging changes. Did you forget to include the pkgproj for this library?

**Comment by Priya91** [comment]
Added the pkgproj, but with it, i get this build error, I don't know how to resolve this..

```
"E:\corefx\src\System.Net.WebSockets.Protocol\pkg\System.Net.WebSockets.Protocol.pkgproj" (default target) (1) ->
(ValidateLibraryPackage target) ->
  E:\corefx\Tools\Packaging.targets(1118,5): error : Framework .NETStandard,Version=v2.0 should support System.Net.WebSockets.Protocol inbox but contained a reference assemblies: ref/netstandard2.0/System.Net.WebSockets.Protocol.dll, ref/netstandard2.0/_._.  You may need to add <InboxOnTargetFramework Include="netstandard2.0" /> to your projec
t. [E:\corefx\src\System.Net.WebSockets.Protocol\pkg\System.Net.WebSockets.Protocol.pkgproj]
  E:\corefx\Tools\Packaging.targets(1118,5): error : Framework .NETStandard,Version=v2.0 should support System.Net.WebSockets.Protocol inbox but contained a implementation assemblies: lib/netstandard2.0/System.Net.WebSockets.Protocol.dll, lib/netstandard2.0/_._.  You may need to add <InboxOnTargetFramework Include="netstandard2.0" /> to your p
roject. [E:\corefx\src\System.Net.WebSockets.Protocol\pkg\System.Net.WebSockets.Protocol.pkgproj]
```

**Comment by Priya91** [comment]
@terrajobst @karelz 

Should the assembly name be system.net.websockets.websocketprotocol, instead of system.net.websockets.protocol? I remember during RuntimeInformation, there was a comment from api review that the assembly name should contain the most common class name from the assembly.

File: src/System.Net.WebSockets.Protocol/pkg/System.Net.WebSockets.Protocol.pkgproj [hunk 2]
```diff
@@ -0,0 +1,12 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
+  <ItemGroup>
+    <ProjectReference Include="..\ref\System.Net.WebSockets.Protocol.csproj">
+      <SupportedFramework>net461;netcoreapp2.0;$(AllXamarinFrameworks)</SupportedFramework>
+    </ProjectReference>
+    <ProjectReference Include="..\src\System.Net.WebSockets.Protocol.csproj" />
+    <InboxOnTargetFramework Include="netstandard2.0" />
```
Comment (weshaggard) [inline src/System.Net.WebSockets.Protocol/pkg/System.Net.WebSockets.Protocol.pkgproj]: Remove this it is what is causing your error.

**Comment by karelz** [comment]
Given that it is one off specific-purpose NuGet package (likely for just this one type with one method ever), I think naming it System.Net.WebSockets.WebSocketProtocol would be fine - @weshaggard @terrajobst?

**Comment by weshaggard** [comment]
>Given that it is one off specific-purpose NuGet package (likely for just this one type with one method ever), I think naming it System.Net.WebSockets.WebSocketProtocol would be fine - @weshaggard @terrajobst?

I agree.

**Comment by Priya91** [comment]
@weshaggard Even without the inboxontargetframework like the build error is same, from the packaging CI run,

```
10:34:40 D:\j\workspace\windows-TGrou---0d2c9ac4\Tools\Packaging.targets(1118,5): error : Framework .NETStandard,Version=v2.0 should support System.Net.WebSockets.WebSocketProtocol inbox but contained a reference assemblies: ref/netstandard2.0/System.Net.WebSockets.WebSocketProtocol.dll.  You may need to add <InboxOnTargetFramework Include="netstandard2.0" /> to your project. [D:\j\workspace\windows-TGrou---0d2c9ac4\src\System.Net.WebSockets.WebSocketProtocol\pkg\System.Net.WebSockets.WebSocketProtocol.pkgproj]
10:34:40 D:\j\workspace\windows-TGrou---0d2c9ac4\Tools\Packaging.targets(1118,5): error : Framework .NETStandard,Version=v2.0 should support System.Net.WebSockets.WebSocketProtocol inbox but contained a implementation assemblies: lib/netstandard2.0/System.Net.WebSockets.WebSocketProtocol.dll.  You may need to add <InboxOnTargetFramework Include="netstandard2.0" /> to your project. [D:\j\workspace\windows-TGrou---0d2c9ac4\src\System.Net.WebSockets.WebSocketProtocol\pkg\System.Net.WebSockets.WebSocketProtocol.pkgproj]
```

**Comment by weshaggard** [comment]
I missed this https://github.com/dotnet/corefx/pull/26250/files#diff-122916076db7087dbc454352fada61eeR2889. You also need to remove that.

File: pkg/Microsoft.Private.PackageBaseline/packageIndex.json [hunk 3]
```diff
@@ -2884,6 +2884,14 @@
         "4.0.1.0": "4.3.0"
       }
     },
+    "System.Net.WebSockets.WebSocketProtocol": {
+      "InboxOn": {
+        "netstandard2.0": "4.0.0.0"
```
Comment (weshaggard) [inline pkg/Microsoft.Private.PackageBaseline/packageIndex.json]: remove this InboxOn section.

File: src/Common/src/System/Net/WebSockets/ManagedWebSocket.common.cs [hunk 4]
```diff
@@ -26,22 +25,8 @@ namespace System.Net.WebSockets
...
-        /// <summary>Creates a <see cref="ManagedWebSocket"/> from a <see cref="Stream"/> connected to a websocket endpoint.</summary>
-        /// <param name="stream">The connected Stream.</param>
-        /// <param name="isServer">true if this is the server-side of the connection; false if this is the client-side of the connection.</param>
-        /// <param name="subprotocol">The agreed upon subprotocol for the connection.</param>
-        /// <param name="keepAliveInterval">The interval to use for keep-alive pings.</param>
-        /// <param name="receiveBufferSize">The buffer size to use for received data.</param>
-        /// <param name="receiveBuffer">Optional buffer to use for receives.</param>
-        /// <returns>The created <see cref="ManagedWebSocket"/> instance.</returns>
-        public static ManagedWebSocket CreateFromConnectedStream(
-            Stream stream, bool isServer, string subprotocol, TimeSpan keepAliveInterval, Memory<byte> buffer)
```
Comment (stephentoub) [inline src/Common/src/System/Net/WebSockets/ManagedWebSocket.common.cs]: I understand why you've made the split you've made, but there's a ton of duplication here.  I don't think it's necessary, and it's going to make it harder to evolve the implementation.

I would prefer to see as minimal changes made as possible, and I don't think a ton should be needed.  For example:
- The new overrides can be moved to a separate .netcoreapp.cs file.
- The netstandard implementation can reference System.Memory so that it can use `Memory<T>`, `Span<T>`, etc.
- Very few new APIs are actually used; in places where they are, you can abstract those away with extension methods that on netcoreapp just target the new methods and on netstandard have a polyfill implementation, e.g. Task.IsCompletedSuccessfully can just do Task.Status == TaskStatus.RanToCompletion, Stream.WriteAsync(Memory) can extra the array and call Stream.WriteAsync(byte[], int, int), etc.

**Comment by stephentoub** [comment]
> I checked for Span on apicatalog, and it said it's available only on .net core

In the box, yes, but it's available from the System.Memory package out-of-band for netstandard / downlevel / etc.

**Comment by Priya91** [comment]
> In the box, yes, but it's available from the System.Memory package out-of-band for netstandard / downlevel / etc.

Oh ok, that makes this change so much easier, I had to split to remove all memory, span stuff, it was painful. Will refactor again.

**Comment by stephentoub** [comment]
> that makes this change so much easier

:smile:

File: src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs [hunk 5]
```diff
@@ -545,7 +522,7 @@ private void SendKeepAliveFrameAsync()
                 Task t = SendFrameLockAcquiredNonCancelableAsync(MessageOpcode.Ping, true, Memory<byte>.Empty);
 
                 // "Observe" any exception, ignoring it to prevent the unobserved exception event from being raised.
-                if (!t.IsCompletedSuccessfully)
+                if (t != null && t.Status != TaskStatus.RanToCompletion)
```
Comment (stephentoub) [inline src/Common/src/System/Net/WebSockets/ManagedWebSocket.cs]: Why do we need to check t != null?

File: src/System.Net.WebSockets.WebSocketProtocol/src/Configurations.props [hunk 6]
```diff
@@ -0,0 +1,8 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <BuildConfigurations>
+      netstandard;
```
Comment (stephentoub) [inline src/System.Net.WebSockets.WebSocketProtocol/src/Configurations.props]: Why aren't we building a netcoreapp config?

File: src/System.Net.WebSockets.WebSocketProtocol/src/System.Net.WebSockets.WebSocketProtocol.csproj [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    </Compile>
+    <Compile Include="$(CommonPath)\System\Net\WebSockets\ManagedWebSocket.cs">
+      <Link>Common\System\Net\WebSockets\ManagedWebSocket.cs</Link>
+    </Compile>
+    <Compile Include="System\Net\WebSockets\ManagedWebSocket.netstandard.cs" />
+    <Compile Include="System\Net\WebSockets\WebSocketProtocol.cs" />
+    <Compile Include="System\Net\WebSockets\ManagedWebSocketExtensions.cs" />
+  </ItemGroup>
+  <ItemGroup>
+    <Reference Include="System.Memory" />
```
Comment (stephentoub) [inline src/System.Net.WebSockets.WebSocketProtocol/src/System.Net.WebSockets.WebSocketProtocol.csproj]: It'd be good if these were alphabetically ordered.

File: src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocketExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.WebSockets
+{
+    internal static class ManagedWebSocketExtensions
+    {
+        internal static string GetString(this UTF8Encoding encoding, Span<byte> bytes)
+        {
+            return encoding.GetString(bytes.ToArray());
```
Comment (stephentoub) [inline src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocketExtensions.cs]: This is super expensive.  It should use the `GetString` overload that takes `byte*`/`int` and use GetReference and fixed to get such a pointer.

File: src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocketExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        }
+
+        internal static Task<int> ReadAsync(this Stream stream, Memory<byte> destination, CancellationToken cancellationToken)
+        {
+            if (destination.TryGetArray(out ArraySegment<byte> array))
```
Comment (stephentoub) [inline src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocketExtensions.cs]: With the call sites this has in the netstandard build, will TryGetArray ever return false?  Seems like the `Memory<byte>` will have always been constructed from a byte[], in which case TryGetArray will always return true, in which case this whole implementation can be:
```C#
bool success = destination.TryGetArray(out ArraySegment<byte> array);
Debug.Assert(success);
return stream.ReadAsync(array.Array, array.Offset, array.Count, cancellationToken);
```

File: src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocketExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+                }
+            }
+        }
+    }
+
+    internal static class BitConverter
+    {
+        internal static int ToInt32(ReadOnlySpan<byte> value)
+        {
+            return System.BitConverter.ToInt32(value.ToArray(), 0);
```
Comment (stephentoub) [inline src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocketExtensions.cs]: Same as earlier comment, this is very expensive; we should not allocate an array here.  You should be able to just copy the implementation from https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/BitConverter.cs#L268-L273.

File: src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/WebSocketProtocol.cs [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System.IO;
+using System.Threading;
+
+namespace System.Net.WebSockets
+{
+    public static class WebSocketProtocol
+    {
+        public static WebSocket CreateFromConnectedStream(
```
Comment (stephentoub) [inline src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/WebSocketProtocol.cs]: Shouldn't this be called `CreateFromStream`?

File: src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/WebSocketProtocol.cs [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+            Stream stream,
+            bool isServer,
+            string subProtocol,
+            TimeSpan keepAliveInterval,
+            ArraySegment<byte> buffer)
```
Comment (stephentoub) [inline src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/WebSocketProtocol.cs]: This should have the same signature as the method on WebSocket:
```C#
Stream stream, bool isServer, string subProtocol, TimeSpan keepAliveInterval, Memory<byte> buffer = default
```
That means the last parameter should be a `Memory<byte>` instead of an `ArraySegment<byte>`, and should have a default value.

File: src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocket.netstandard.cs [hunk 13]
```diff
@@ -0,0 +1,26 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+
+// NOTE: This file is shared between CoreFX and ASP.NET.  Be very thoughtful when changing it.
```
Comment (stephentoub) [inline src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocket.netstandard.cs]: This comment and the others like it can be removed.  The purpose of this change is to remove the need for such code sharing.

File: src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocket.netstandard.cs [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+{
+    internal sealed partial class ManagedWebSocket : WebSocket
+    {
+        private Task ValidateAndReceiveAsync(Task receiveTask, byte[] buffer, CancellationToken cancellationToken)
+        {
+            if (receiveTask == null ||
+                (receiveTask.Status == TaskStatus.RanToCompletion &&
+                !(receiveTask is Task<WebSocketReceiveResult> wsrr && wsrr.Result.MessageType == WebSocketMessageType.Close)))
+            {
+                receiveTask = ReceiveAsyncPrivate<WebSocketReceiveResultGetter, WebSocketReceiveResult>(new ArraySegment<byte>(buffer), cancellationToken).AsTask();
```
Comment (stephentoub) [inline src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocket.netstandard.cs]: Why did you remove the special-casing of ReceiveAsyncPrivate returning an already successfully completed task?

Comment (stephentoub) [reply src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocket.netstandard.cs]: Nevermind; I was looking at the wrong method.

File: src/System.Net.WebSockets.WebSocketProtocol/tests/WebSocketProtocolTests.cs [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+using System.Net.Http;
+using System.Net.Sockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Net.WebSockets.Tests
+{
+    public sealed class WebSocketProtocolTests
```
Comment (stephentoub) [inline src/System.Net.WebSockets.WebSocketProtocol/tests/WebSocketProtocolTests.cs]: How are we testing the WebSocket returned from WebSocketProtocol.CreateFromStream?  The only test I see below that exercises it is WebSocketProtocol_CreateFromConnectedStream_Succeeds, which tests very little of it, e.g. it doesn't call ReceiveAsync at all.

File: src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.netcoreapp.cs [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+            catch (Exception exc)
+            {
+                return new ValueTask<ValueWebSocketReceiveResult>(Task.FromException<ValueWebSocketReceiveResult>(exc));
+            }
+        }
+
+        private Task ValidateAndReceiveAsync(Task receiveTask, byte[] buffer, CancellationToken cancellationToken)
+        {
+            if (receiveTask == null ||
+                        (receiveTask.Status == TaskStatus.RanToCompletion &&
```
Comment (stephentoub) [inline src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.netcoreapp.cs]: This can be changed from:
```C#
receiveTask.Status == TaskStatus.RanToCompletion
```
to
```C#
receiveTask.IsCompletedSuccessfully
```
They're functionally identical but the latter is more efficient.

File: src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocketExtensions.cs (hunk 9 above)
Comment (Priya91) [reply src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocketExtensions.cs]: The buffer passed may come from user, from the callsite in managedsocket, the buffer passed here could be the memory<byte> from WebSocketProtocol.CreateFromStream api.

File: src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.netcoreapp.cs (hunk 16 above)
Comment (Priya91) [reply src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.netcoreapp.cs]: Looking at the property implementation, it looks like it is doing the same check, how is it efficient?

File: src/System.Net.WebSockets.WebSocketProtocol/tests/WebSocketProtocolTests.cs (hunk 15 above)
Comment (Priya91) [reply src/System.Net.WebSockets.WebSocketProtocol/tests/WebSocketProtocolTests.cs]: I had mentioned this in the issue post, re-commenting here, 

I was not able to get a successful test with WebSocket.ReceiveAsync, it fails in ReceiveAsync, because the response from the server is not in the right format. So the code fails with ProtocolError in ReceiveAsync. The behavior is same with and without this refactoring. Hence didn't include ReceiveAsync in the WebSocketProtocol_CreateFromConnectedStream_Succeeds outerloop test.

I can include this as a skipped test, to unblock this PR, and file a follow up issue to investigate the failure.

File: src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.netcoreapp.cs (hunk 16 above)
Comment (stephentoub) [reply src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.netcoreapp.cs]: What implementations are you comparing?

File: src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocketExtensions.cs (hunk 9 above)
Comment (stephentoub) [reply src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocketExtensions.cs]: Ok

File: src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.netcoreapp.cs (hunk 16 above)
Comment (Priya91) [reply src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.netcoreapp.cs]: Was looking at this, https://github.com/dotnet/corefx/blob/master/src/System.Threading.Tasks.Extensions/src/System/Threading/Tasks/ValueTask.cs#L118

File: src/System.Net.WebSockets.WebSocketProtocol/ref/System.Net.WebSockets.WebSocketProtocol.csproj [hunk 17]
```diff
@@ -0,0 +1,16 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'netstandard-Debug|AnyCPU'" />
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'netstandard-Release|AnyCPU'" />
+  <ItemGroup>
+    <Compile Include="System.Net.WebSockets.WebSocketProtocol.cs" />
+  </ItemGroup>
+  <ItemGroup>
+    <Reference Include="System.IO" />
```
Comment (weshaggard) [inline src/System.Net.WebSockets.WebSocketProtocol/ref/System.Net.WebSockets.WebSocketProtocol.csproj]: All these need to be ProjectReference

File: src/System.Net.WebSockets.WebSocketProtocol/tests/WebSocketProtocolTests.cs [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+        [OuterLoop] // Connects to external server.
+        [Fact]
+        [ActiveIssue(26288)]
+        public async Task WebSocketProtocol_CreateFromConnectedStream_Succeeds()
+        {
+            Socket s = new Socket(SocketType.Stream, ProtocolType.Tcp);
+            await s.ConnectAsync("corefx-net.cloudapp.net", 80);
+            using (NetworkStream ns = new NetworkStream(s))
+            {
+                using (WebSocket socket = WebSocketProtocol.CreateFromStream(ns, false, null, TimeSpan.FromSeconds(10)))
```
Comment (stephentoub) [inline src/System.Net.WebSockets.WebSocketProtocol/tests/WebSocketProtocolTests.cs]: This is buggy.  As mentioned in https://github.com/dotnet/corefx/pull/26296/files#r161290252, CreateFromStream is meant to be used with a stream that's already connected for websocket communication; this isn't.

File: src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.netcoreapp.cs (hunk 16 above)
Comment (stephentoub) [reply src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.netcoreapp.cs]: That's ValueTask.  This is Task.

Comment (stephentoub) [reply src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.netcoreapp.cs]: And that's also the corefx implementation of ValueTask, which works downlevel.  Compare it to the coreclr implementation: https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/Threading/Tasks/ValueTask.cs#L132

Comment (Priya91) [reply src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.netcoreapp.cs]: I still don't get why it's more efficient though..

File: src/System.Net.WebSockets.WebSocketProtocol/tests/WebSocketProtocolTests.cs (hunk 18 above)
Comment (stephentoub) [reply src/System.Net.WebSockets.WebSocketProtocol/tests/WebSocketProtocolTests.cs]: This also isn't connecting to an echo WebSocket, so even if the connection was upgraded, the receive would still hang as there's nothing to receive.

File: src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.netcoreapp.cs (hunk 16 above)
Comment (stephentoub) [reply src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.netcoreapp.cs]: > I still don't get why it's more efficient though

Have you looked at the implementations?
https://github.com/dotnet/coreclr/blob/140ba024bf6edb61744fe6b84bd635d389fe1413/src/mscorlib/src/System/Threading/Tasks/Task.cs#L1458
vs
https://github.com/dotnet/coreclr/blob/140ba024bf6edb61744fe6b84bd635d389fe1413/src/mscorlib/src/System/Threading/Tasks/Task.cs#L1329-L1351

File: src/System.Net.WebSockets.WebSocketProtocol/tests/WebSocketProtocolTests.cs (hunk 18 above)
Comment (Priya91) [reply src/System.Net.WebSockets.WebSocketProtocol/tests/WebSocketProtocolTests.cs]: Yeah changing the test.

**Comment by Priya91** [comment]
@dotnet-bot test outerloop windows x64 debug build please
@dotnet-bot test outerloop linux x64 debug build please
@dotnet-bot test outerloop osx x64 debug build please

**Comment by Priya91** [comment]
@stephentoub I've addressed the PR feedback, can you take a quick look..

File: src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocketExtensions.cs [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+            }
+        }
+    }
+
+    internal static class BitConverter
+    {
+        internal static unsafe int ToInt32(ReadOnlySpan<byte> value)
+        {
+            if (value.Length < sizeof(int))
+                throw new ArgumentOutOfRangeException(nameof(value));
```
Comment (stephentoub) [inline src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocketExtensions.cs]: Based on the call sites, will this ever fail? Wondering if it can instead be an assert.

Comment (Priya91) [reply src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocketExtensions.cs]: The function that calls to the BitConverter takes a byte array, and slices it based on an offset. It could be the case that length-offset < sizeof(int).. The byte array is _sendBuffer which is atleast 14 bytes in length, and markOffset comes from WriteHeader.

Comment (stephentoub) [reply src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocketExtensions.cs]: > It could be the case that length-offset < sizeof(int)

When? You're saying there are reachable cases today where this results exceptions being thrown?

Comment (Priya91) [reply src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocketExtensions.cs]: I am not sure, it depends upon what the runtime values of markOffset and _sendBuffer could be..

Comment (stephentoub) [reply src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocketExtensions.cs]: I'm suggesting we validate that indeed the inputs will always be valid and then change this to an assert.  If this did throw an exception, it indicates a problem in earlier logic that should have guaranteed there was enough data for this to be used.

Comment (Priya91) [reply src/System.Net.WebSockets.WebSocketProtocol/src/System/Net/WebSockets/ManagedWebSocketExtensions.cs]: There is a debug.assert in applymask that checks the conditions of the bytearray and maskoffset before calling the method that calls this code, so this can be changed to assert.

**Comment by Priya91** [comment]
@dotnet-bot test Linux x64 Release Build please
@dotnet-bot test UWP CoreCLR x64 Debug Build please
@dotnet-bot test Windows x64 Debug Build please
@dotnet-bot test Windows x86 Release Build please

@dotnet-bot test outerloop windows x64 debug build please
@dotnet-bot test outerloop linux x64 debug build please
@dotnet-bot test outerloop osx x64 debug build please

**Comment by Priya91** [comment]
I've addressed the PR feedback, and the CI passed. Merging this PR to make progress. If there are any other concerns, it can be addressed separately.
